File: src/cleansia_android/partner-app/src/main/java/cz/cleansia/partner/features/profile/CountryLocalization.kt
package cz.cleansia.partner.features.profile

import cz.cleansia.core.settings.SupportedLanguages
import cz.cleansia.partner.api.model.CountryListItem
import java.util.Locale

/**
 * The country's name in the reader's language.
 *
 * `CountryListItem.name` is the raw `Countries.Name` column, which is English by definition — the seed
 * stores every other language under `Translations`. Reading `.name` directly, which every Android call
 * site did, showed a Czech partner "Czech Republic" in an otherwise fully Czech screen. The web has
 * always resolved it this way (`country.translations?.[currentLang]?.name` in the profile facades); this
 * is the mobile side of the same rule.
 *
 * The lookup goes through [SupportedLanguages.bareCode] because `Locale.getDefault().toLanguageTag()`
 * hands back `"cs-CZ"` on a Czech handset while the translation map is keyed by bare `"cs"` — matching
 * without the narrowing step silently never hits, which is the failure mode this replaces rather than
 * the one it introduces.
 *
 * Falls back through name → ISO code → id, so a country seeded without a translation for this language
 * still renders something a person can read.
 */
fun CountryListItem.localizedName(locale: Locale = Locale.getDefault()): String {
    val code = SupportedLanguages.bareCode(locale.toLanguageTag())
    val translated = code
        ?.let { translations?.get(it) }
        ?.name
        ?.takeIf { it.isNotBlank() }

    return translated
        ?: name?.takeIf { it.isNotBlank() }
        ?: isoCode?.takeIf { it.isNotBlank() }
        ?: id.orEmpty()
}

File: src/cleansia_ios/CleansiaPartner/Sources/Features/Profile/CountryLocalization.swift
import CleansiaCore
import CleansiaPartnerApi
import Foundation

public extension CountryListItem {
    /// The country's name in the reader's language.
    ///
    /// `CountryListItem.name` is the raw `Countries.Name` column, which is English by definition — the
    /// seed stores every other language under `translations`. Reading `.name` directly, which every
    /// iOS call site did, showed a Czech partner "Czech Republic" in an otherwise fully Czech screen.
    /// The web has always resolved it this way (`country.translations?.[currentLang]?.name` in the
    /// profile facades); this is the mobile side of the same rule.
    ///
    /// **Keyed on the language the reader CHOSE, not on the device's.** This defaulted to
    /// `Locale.current`, and the in-app language switch never reaches that: it repoints the string
    /// bundles but does not set `AppleLanguages`, so the process locale reports the device language
    /// forever. A partner running the app in Czech on an English phone got "Czech Republic" in an
    /// otherwise Czech screen — the exact bug this method was written to fix, reintroduced one layer
    /// down by trusting the wrong source. `CoreL10n.languageTag` is what the switch actually sets.
    ///
    /// The lookup still narrows through `UserDefaultsAppSettingsStore.bareLanguageCode` because a tag
    /// may be region-qualified (`"cs-CZ"`) while the translation map is keyed by bare `"cs"` —
    /// matching without that step silently never hits.
    ///
    /// Falls back through name → ISO code → id, so a country seeded without a translation for this
    /// language still renders something a person can read.
    ///
    /// Twin: `CountryListItem.localizedName` (Android, partner-app). Keep the two in step.
    func localizedName(languageTag: String = CoreL10n.languageTag) -> String {
        let translated = UserDefaultsAppSettingsStore.bareLanguageCode(languageTag)
            .flatMap { translations?[$0]?.name }
            .flatMap { $0.isBlank ? nil : $0 }

        return translated
            ?? name.flatMap { $0.isBlank ? nil : $0 }
            ?? isoCode.flatMap { $0.isBlank ? nil : $0 }
            ?? id
            ?? ""
    }
}
